crates/sdl-encoder/src/union_def.rs

```diff
@@ -0,0 +1,106 @@
+use std::fmt::{self, Display};
+
+/// UnionDefs are an abstract type where no common fields are declared.
+///
+/// *UnionDefTypeDefinition*:
+///     Description<sub>opt</sub> **union** Name Directives<sub>\[Const\] opt</sub> UnionDefMemberTypes<sub>opt</sub>
+///
+/// Detailed documentation can be found in [GraphQL spec](https://spec.graphql.org/draft/#sec-UnionDef).
+///
+/// ### Example
+/// ```rust
+/// use sdl_encoder::{UnionDef};
+///
+/// let mut union_ = UnionDef::new("Cat".to_string());
+/// union_.description(Some(
+///     "A union of all cats represented within a household.".to_string(),
+/// ));
+/// union_.member("NORI".to_string());
+/// union_.member("CHASHU".to_string());
+///
+/// assert_eq!(
+///     union_.to_string(),
+///     r#""""A union of all cats represented within a household."""
+/// union Cat = NORI | CHASHU
+/// "#
+/// );
+/// ```
+#[derive(Debug, PartialEq, Clone)]
+pub struct UnionDef {
+    // Name must return a String.
+    name: String,
+    // Description may return a String.
+    description: Option<String>,
+    // The vector of members that can be represented within this union.
+    members: Vec<String>,
```

Comment:
It might be fine to represent these as strings for now if all we use it for is SDL output, but they really are references to object types, so these are type names.

Reply:
This is an encoder, so it will only ever handle strings to generate a string representation of a schema. It is not the job of a serialiser/encoder to keep references to the object type, that is done in the AST.

Reply:
Ah, I think that is why some of my suggestions may not make as much sense to you. I was thinking of this as a schema definition layer similar to `GraphQLSchema` in `graphql-js`. We still need that if we want to do anything beyond encoding, but I can see how that isn't a priority right now. 

Note that the AST doesn't keep references to types either, it just contains names as strings (and also fails to contain other meaningful relationships). So I do think we need a layer of schema and type definitions as part of our `apollo-rs` library. At that point, it seems the types in `sdl-encoder` would be superfluous and maybe we could find other ways to add encoding behavior to a schema definition layer (by implementing encoding traits for types defined elsewhere for example).

Reply:
We will eventually have a schema definition layer you're talking about based on what's in `crates/rover-client/introspection/schema.rs`. Meaning it will encode its internal representation of the schema the way it does currently when we need an SDL. The current encoder is still missing some things to be able to do non-introspection based SDL serialisations, but that's something we will add later.

In general, I think discussing architecture of the future graphql rust tooling is out of scope for this PR review. I'd prefer to have that discussion separately as we carry out the planning and design steps for it.